notes.md

```diff
@@ -0,0 +1,215 @@
+# Why does `:` only sometimes work?
+I don't like `:` being listed as a `source_column_selector` when it often errors.
+
+`select(df, :)`
+: works
+
+`transform(df, Cols(:) => ByRow(max))`
+: works
+
+`transform(df, : => ByRow(max))` errors
+: ERROR: syntax: whitespace not allowed after ":" used for quoting
+
+Similarly, regular expressions also sometimes must be wrapped in `Cols`.
+This seems inconsistent.
+I don't understand the rules for when `Cols` is and isn't needed.
+It should be explained in the documentation.
+
+```julia
+julia> df = DataFrame(Time=[3, 4, 5], TopTemp=[70, 73, 100], BottomTemp=[50, 55, 80])
+3×3 DataFrame
+ Row │ Time   TopTemp  BottomTemp
+     │ Int64  Int64    Int64
+─────┼────────────────────────────
+   1 │     3       70          50
+   2 │     4       73          55
+   3 │     5      100          80
+
+julia> transform(df, Cols(r"Temp") .=> (t->(t.-32)*5/9), renamecols=false)
+3×3 DataFrame
+ Row │ Time   TopTemp  BottomTemp
+     │ Int64  Float64  Float64
+─────┼────────────────────────────
+   1 │     3  21.1111     10.0
+   2 │     4  22.7778     12.7778
+   3 │     5  37.7778     26.6667
+
+julia> transform(df, r"Temp" .=> (t->(t.-32)*5/9), renamecols=false) # breaks without Cols
```

Comment:
`r"Temp"` lists columns AFTER broadcasting.

`Cols(r"Temp")` lists columns WITHIN broadcasting.

Reply:
It looks like the same not-expanded-yet issue occurs with `(:)`, but then I'm confused why `2:3` works fine. Shouldn't the `2:3` behavior match the `(:)` behavior?
```julia
julia> df = DataFrame(a = 1:3, b = 4:6, c = 7:9)
3×3 DataFrame
 Row │ a      b      c     
     │ Int64  Int64  Int64
─────┼─────────────────────
   1 │     1      4      7
   2 │     2      5      8
   3 │     3      6      9

julia> transform(df, (:) .=> ByRow(x -> x + 1))
ERROR: MethodError: no method matching (::var"#23#24")(::Int64, ::Int64, ::Int64)

julia> transform(df, Cols(:) .=> ByRow(x -> x + 1))
3×6 DataFrame
 Row │ a      b      c      a_function  b_function  c_function 
     │ Int64  Int64  Int64  Int64       Int64       Int64
─────┼─────────────────────────────────────────────────────────
   1 │     1      4      7           2           5           8
   2 │     2      5      8           3           6           9
   3 │     3      6      9           4           7          10

julia> transform(df, 2:3 .=> ByRow(x -> x + 1))
3×5 DataFrame
 Row │ a      b      c      b_function  c_function 
     │ Int64  Int64  Int64  Int64       Int64
─────┼─────────────────────────────────────────────
   1 │     1      4      7           5           8
   2 │     2      5      8           6           9
   3 │     3      6      9           7          10
```

Reply:
`(:)` cannot be broadcasted for the same reason as regex - it is defined in Base.

There is no problem with `2:3` because it explicitly specifies which columns to pick (you do not need to know `df` to determine that). While `:` or regrex are dynamic - they can be resolved only in the context of `df` to know what columns are picked. That is why `Cols` wrapper is needed.